pipeline/antispam/antispammer.go

```diff
@@ -0,0 +1,188 @@
+package antispam
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ozontech/file.d/logger"
+	"github.com/ozontech/file.d/metric"
+	prom "github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// Antispammer makes a decision on the need to parse the input log.
+// It can be useful when any application writes logs at speed faster than File.d can read it.
+//
+// Anti-spammer supports exceptions for cases where you need to guarantee delivery of an important events.
+type Antispammer struct {
+	unbanIterations int
+	threshold       int
+	mu              sync.RWMutex
+	sources         map[uint64]source
+	exceptions      []Exception
+
+	logger *zap.Logger
+
+	// antispammer metrics
+	activeMetric    *prom.GaugeVec
+	banMetric       *prom.GaugeVec
+	exceptionMetric *prom.CounterVec
+}
+
+type source struct {
+	counter atomic.Int32
+	name    string
+}
+
+type Options struct {
+	MaintenanceInterval time.Duration
+	Threshold           int
+	UnbanIterations     int
+	Exceptions          []Exception
+
+	Logger            *zap.Logger
+	MetricsController *metric.Ctl
+}
+
+func NewAntispammer(o Options) *Antispammer {
+	if o.Threshold != 0 {
```

Comment:
I think we should forbid using `o.Threshold` if it is less than zero, or pass to `Antispammer.threshold` zero in that case. It is unlikely that someone on purpose will write negative value to `antispam_threshold` in config, but still this behaviour should be considered.

Reply:
done